bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/obs/plannedtime/PlannedTime.java

```diff
@@ -223,17 +222,30 @@ public static CategorizedTimeGroup apply(CategorizedTime... col) {
             return times.hashCode();
         }
 
-        public Map<Category, CategorizedTime> maxTimes() {
+        public Pair<Map<Category, CategorizedTime>, Map<Category, ImList<CategorizedTime>>> maxTimes() {
```

Comment:
`maxTimes` is not quite accurate anymore since we're returning everything, just partitioned. Maybe a way to simplify would be to "groupBy" `Category`, returning just one `Map<Category, ImList<CategorizedTime>>` with everything in it (no finding the max here).  You could call it `groupTimes` or something.

Then in the `ObserveTimeLineNode` where you need to pick out the max, you can use `ImList`'s `sort` method to find it.  i.e.

```java
  final ImList<CategorizedTime> s = catTimes.sort(Comparator.naturalOrder()); // I think ...
  final CategorizedTime max = s.head();
  final ImList<CategorizedTime> losers = s.tail();
```

Reply:
Well, except that I suppose you'd be picking the min that way so it'd need to be reversed or you'd have to get the `last` and `init` instead of `head` and `tail`!

Reply:
So you can do all that with `ImList`s 😮? That's great.

I'm realizing `ImList` provides a very decent functional API. I guess I might be underestimating Java as a capable FP language.

Reply:
Thinking about it more, you don't really need to sort them, there's also `max`:

```
final CategorizedTime max = catTimes.max(Comparator.naturalOrder());
```

You want all the items including the max anyway for the sub-table.

There is no `groupBy` in `ImList` but I guess we could add it for this.